### ISSUE TITLE
Variable closure should be supported by GO_OP

### ISSUE THREAD

**Issue by cs2be** [PaddlePaddle/Paddle#8502]
@reyoung While we were implementing the GO_OP, we realized that we may need to provide block captures (ie: closures) for scope variables.  Please see below for details...

### Intro

Fluids allows variables to exists within scopes.  Variables can be persistable (exists in global scope) or non-persistable .  When a sub block is added to the main block, a new child local scope is created from the parent scope.  The sub block is able to access variables from within its local and parent scopes.

This works well for a single threaded case (since the parent blocks will never exit before the child blocks), however we run into problems when we introduce a multi-threaded scenario with the GO_OP.

GO_OP creates a new block, operators and local variables.  It then executes the block in a new thread.  Since the block executes asynchronously from the parent block, the parent block could potentially have exited while the GO_OP is executing.  

In our current architecture, when the parent block exits, it deletes it's local scope, any child scopes, and any local variables that in that local scope.  If the GO_OP creates a local scope from the parent block's local scope, then this will cause an issue when the GO_OP operators tries to access any parent scope variables.

Take the example below:

```
cond = layers.less_than(x=a0, y=a1)
ie = layers.IfElse(cond)
      
with ie.true_block():
  a0 = layers.data(name='a0', shape=[1], dtype='int64')
  a1 = layers.data(name='a1', shape=[1], dtype='int64')
  
  with fluid.Go():
    # This will cause an error.  a0/a1 exists in the ie.true_block local scope.
    # When fluid.Go() executes, ie.true_block may have already exited
    a_sum = layers.sums(input=[mean_a0, mean_a1])
```

Before the fluid.Go() operator is ran, the scopes looks like this:

![fluidclosure1](https://user-images.githubusercontent.com/403396/36566840-79d26a5a-17d9-11e8-8a6b-cc3cee252dea.png)

After fluid.Go() operator spins off a new thread, the ie.true_block() will complete and destroy any child scopes and variabless:

![fluidclosure2](https://user-images.githubusercontent.com/403396/36566855-80d06d20-17d9-11e8-83ac-74f31f2f4a22.png)

To get around the issue of the scopes being deleted, the fluid.Go operator can create a new scope off the global scope, and add a shared_ptr to any parent variables it needs.  That way when the parent scope gets deleted, the child fluid.Go scope does not get removed.  In addition, if we use a shared pointer to a0 and a1 object, then the object will not be deleted either.

![fluidclosure3](https://user-images.githubusercontent.com/403396/36566859-87f2b298-17d9-11e8-8634-ac5b2839bb4a.png)

An alternate design could be we create a new fluid.Go scope that has no parents, and it just "block captures" any variable it needs.

![fluidclosure4](https://user-images.githubusercontent.com/403396/36566868-909aeffa-17d9-11e8-82d2-066fbb9f2462.png)

This change will require us to update scope.cc to have it use std::shared_ptr<Variable> references, instead of Variable* references.

Please see the gist of how this maybe done.

https://gist.github.com/cs2be/2fb810bc16631a905dc02e34c6d00e5a
https://gist.github.com/cs2be/50dd2540f1a43445e74ff0f5f1dc8097

Please let us know if this design would work, or if you see any potential issues.  Thanks.

**Comment by reyoung** [comment]
## Problems

Just like [`lambda in C++`](http://en.cppreference.com/w/cpp/language/lambda), our `GO_OP` needs to capture local variables. In the description of this issue, only captures by reference is discussed.

However, there are three kinds of captures in C++, there are:

* Capture by value. (Just copy a new variable into sub-scope).
* Capture by reference. (Just copy the pointer into sub-scope). We can implement this by `shared_ptr`.
* Capture by moving value. i.e. move a variable into sub-scope.

I think the simplest capture is `capture by value`. We can support this feature to verify our `GO_OP` design. The `Channel` could be a `shared_ptr<FIFOQueue<T>>`. We can capture tensors and channels by value. This implementation will be slow since there are a lot of memory copies.  However, we can implement other capture methods when we can use `GO_OP` to train a neural network.

If we make `GO_OP` be able to train a neural network, we can implement `capture by reference` and `capture by moving value` later. 

## Concerns

I am afraid that this issue is too complicated to give a simple solution because to introduce CSP or thread will influence several parts of `Fluid`. They are:

1.  How to `backward` the `GO_OP`? Or should `GO_OP` support backward? 
     * If `GO_OP` supports `backwards`, the Variable closure should take care of gradients.
          * How does `GO_OP` get the error signal through backwards? `GO_OP` will put outputs into a channel. How to get the gradients of these outputs?
        * How does `GO_OP` produce the error signal to outside? The inputs of `GO_OP` is captured. So we should accumulate the inside gradient, which is generated in `GO_OP` backward stage, to the outside gradient? Since the `GO_OP` is asynchronous, how to do that?
     * If `GO_OP` does not support `backward`, we cannot use `GO_OP` to create a neural network, since we can only train a neural network by backwards. We cannot use `GO_OP` implements `Parallel.Do` since `Parallel.Do` supports backward.
     * The fluid backwards is designed and implemented by me and @JiayiFeng . The `Parallel.Do` is designed and implemented by @tonyyang-svail .

2. How to implement memory optimization?
     * The memory optimization is based on static analysis while compiling. Since the variable closure is kind like sharing variables between blocks. The previous memory optimization algorithm, which is based on http://mxnet.incubator.apache.org/architecture/note_memory.html, should be redesigned.
     * The memory optimization is designed and implemented by @QiJune .

**Comment by cs2be** [comment]
@reyoung Thank you for the quick response.  We'll discuss with team if the first iteration of CSP will be used for non-training purposes (ie: Go thread uploading gradients or reading minibatch https://github.com/PaddlePaddle/Paddle/blob/6024a170f321e3ed572260e68cad39820f15ff67/doc/design/csp.md), and later extend CSP to support back prop.

Since we are encapsulating the ChannelHolder into a Variable*, do you mean to make a copy of the channel pointer?  Since the Channel uses a shared lock, we're not sure how to share this lock if we just copy the contents of the Channel.  We can however copy the data coming into and out of the channel.
